lib/RestructureCFG/CycleEquivalenceAnalysis.cpp

```diff
@@ -0,0 +1,659 @@
+  return TargetDFSNum <= SourceDFSNum;
+}
+
+template<class GraphT, class GT>
+void CycleEquivalenceAnalysis<GraphT, GT>::computeDFSAndSpanningTree(BlockGraph
+                                                                       &Graph) {
+  using ExtType = llvm::df_iterator_default_set<BlockNode *>;
+  using UGT = llvm::GraphTraits<llvm::Undirected<BlockNode *>>;
+  using udf_iterator = llvm::df_iterator<BlockNode *, ExtType, false, UGT>;
+
+  BlockNode *EntryNode = Graph.getEntryNode();
+  auto It = udf_iterator::begin(EntryNode);
+  auto End = udf_iterator::end(EntryNode);
+
+  size_t DFSNum = 0;
+  for (; It != End; It++) {
+    BlockNode *CurrentNode = *It;
+
+    // We assign the DFSNum for `CurrentNode`
+    CurrentNode->setDFSNum(DFSNum);
+    DFSNum++;
+
+    // Each time we reach a new node during the `DFS` visit, we search for the
+    // edge that brougth us here, and that we therefore need to insert into the
+    // spanning tree.
+    // Retrieve the last-but-one element on the `VisitStack`, which is the
+    // candidate as the source node of the edge we are interested into adding to
+    // the spanning tree.
+    auto VisitStackSize = It.getPathLength();
+
+    // When we have a single node on the `VisitStack`, we cannot assign an edge
+    // to the spanning tree
+    if (VisitStackSize < 2)
+      continue;
+
+    BlockNode *SourceNode = It.getPath(VisitStackSize - 2);
+
+    // Search for the edge which connects `Source` and `CurrentNode`, which is
+    // the edge composing the tree edge of the current spanning tree exploration
+    using UndirectedGraphT = llvm::Undirected<BlockNode *>;
+    auto
+      ChildrenEdgesRange = llvm::children_edges<UndirectedGraphT>(SourceNode);
+    const auto FindCurrentNode = [&CurrentNode](const auto &Pair) {
+      return Pair.Neighbor == CurrentNode;
+    };
+    auto EdgeIt = llvm::find_if(ChildrenEdgesRange, FindCurrentNode);
+    revng_assert(EdgeIt != ChildrenEdgesRange.end());
+    EdgeIt->Label->setType(EdgeType::TreeEdge);
+  }
+
+  // All the remaining edges that have not been marked as `TreeEdge`s, are the
+  // `BackEdge`s. Mind that we cannot perform this assignment during the
+  // previous visit itself, because we first need to visit all the edges to mark
+  // the `TreeEdge`s
+  for (BlockNode *CurrentNode :
+       llvm::depth_first(llvm::Undirected(Graph.getEntryNode()))) {
+    for (auto &[Neighbor, Label] :
+         llvm::children_edges<llvm::Undirected<BlockNode *>>(CurrentNode)) {
+      if (Label->type() == EdgeType::Invalid) {
+        Label->setType(EdgeType::BackEdge);
+      }
+    }
+  }
+}
+
+/// Helper function to materialize the capping backedges
+template<class GraphT, class GT>
+void CycleEquivalenceAnalysis<GraphT, GT>::insertCappingBackedges(BlockGraph
+                                                                    &Graph) {
+  // Highest, i.e., closer to the spanning tree root, reached DFSNum for each
+  // `BB`
```

Comment:
Please add that for nodes that are "higher", so close to the root, the `DFSNum` is smaller. Otherwise the rest gets confusing. At least if you say it out loud in a comment there's a place to refer to when you're confused, and it doesn't seem wrong to use `std::min` all over the place to compute the "highest".

BTW why did you decide to use the "highest" on your own or to use the same terminology of the paper?

If you did it for the paper, keep it like this but explain that's the reason. It can still get confusing but if you read the paper everything makes sense. 
Instead, if it was your choice please revert it all over the codebase because it can get very confusing for no actual reason.

Reply:
The terminology is taken from the paper, I tried to keep it as coherent as possible with it. I added here https://github.com/revng/revng-c/blob/ce352f2912e639e1ef307bb1d5b131db48444878/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L178 the disclaimer on top on the explanation "highest -> lower dfsnum", let me know if you still want to change the use of the terms.

Reply:
I already had a comment before the `CycleEquivalenceAnalasysis` class [here](https://github.com/revng/revng-c/blob/f23593da9f15288aca2fee41975e6e55dbbbcb3b/include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h#L119), but I also added a small reference, as you suggested, [here](https://github.com/revng/revng-c/blob/f23593da9f15288aca2fee41975e6e55dbbbcb3b/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L5) at the beginning of the `cpp` file.

Reply:
I remember there wasn't one, but I couldn't find it during review. It must have been an off-by-grep. Thanks for adding it anyways.